router/pkg/mcpserver/server.go

```diff
@@ -0,0 +1,676 @@
+	return func(o *Options) {
+		o.EnableArbitraryOperations = enableArbitraryOperations
+	}
+}
+
+// WithExposeSchema sets the expose schema option
+func WithExposeSchema(exposeSchema bool) func(*Options) {
+	return func(o *Options) {
+		o.ExposeSchema = exposeSchema
+	}
+}
+
+// LoadOperations loads operations from the configured operations directory
+func (s *GraphQLSchemaServer) LoadOperations() error {
+	return s.operationsManager.LoadOperationsFromDirectory(s.operationsDir)
+}
+
+// LoadOperationsFromDirectory loads operations from a specified directory
+func (s *GraphQLSchemaServer) LoadOperationsFromDirectory(operationsDir string) error {
+	return s.operationsManager.LoadOperationsFromDirectory(operationsDir)
+}
+
+// ServeSSE starts the server with SSE transport
+func (s *GraphQLSchemaServer) ServeSSE() (*server.SSEServer, *http.Server, error) {
+	listenAddr := "localhost:" + strconv.Itoa(s.port)
+
+	// Create HTTP server with timeouts
+	httpServer := &http.Server{
+		Addr:         listenAddr,
+		ReadTimeout:  15 * time.Second,
+		WriteTimeout: 15 * time.Second,
+		IdleTimeout:  30 * time.Second,
+	}
+
+	sseServer := server.NewSSEServer(s.server,
+		server.WithBaseURL(fmt.Sprintf("http://%s", listenAddr)),
+		server.WithHTTPServer(httpServer),
+		server.WithSSEEndpoint("/mcp"),
+		server.WithSSEContextFunc(authFromRequest),
+	)
+
+	go func() {
+		err := sseServer.Start(listenAddr)
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			s.logger.Error("failed to start SSE server", zap.Error(err))
+		}
+	}()
+
+	return sseServer, httpServer, nil
+}
+
+// Start loads operations and starts the server
+func (s *GraphQLSchemaServer) Start() error {
+
+	if err := s.LoadOperations(); err != nil {
+		return fmt.Errorf("failed to load operations: %w", err)
+	}
+
+	sseServer, httpServer, err := s.ServeSSE()
+	if err != nil {
+		return fmt.Errorf("failed to create SSE server: %w", err)
+	}
+
+	// Store server references for Stop method
+	s.sseServer = sseServer
+	s.httpServer = httpServer
+
+	// Register tools
+	s.registerTools()
+
+	return nil
+}
+
+// Stop gracefully shuts down the MCP server
+func (s *GraphQLSchemaServer) Stop(ctx context.Context) error {
+	if s.httpServer == nil {
+		return nil
+	}
+
+	s.logger.Debug("shutting down MCP server")
+
+	// Create a shutdown context with timeout
+	shutdownCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	// Shutdown the HTTP server
+	if err := s.httpServer.Shutdown(shutdownCtx); err != nil {
+		return fmt.Errorf("failed to gracefully shutdown MCP server: %w", err)
+	}
+
+	return nil
+}
+
+// registerTools registers all tools for the MCP server
+func (s *GraphQLSchemaServer) registerTools() {
+	// TODO: Support tool annotations when available in MCP Go SDK
+
+	// Add a tool to list all available operations (names and descriptions only)
+	s.server.AddTool(
+		mcp.NewTool(
+			"list_operations",
```

Comment:
can you explain the difference between the "list_operations" tool and using MCP introspection to get the schema of all tools?

Reply:
Great question. I removed it. In the beginning, I wasn't sure if the operations are properly available (with description) in the context of the session.